hera_qm/vis_metrics.py

```diff
@@ -48,6 +50,105 @@ def check_noise_variance(data):
     return Cij
 
 
+def sequential_diff(data, wgts=None, axis=(0,)):
+    """
+    Take a sequential (adjacent) difference of a visibility waterfall
+    as an estimate of the visibility noise.
+
+    Parameters
+    ----------
+    data : ndarray or UVData object
+        A 2D or 3D ndarray containing visibility data, with
+        shape (Ntimes, Nfreqs, :). Or UVData object
+        holding visibility data.
+
+    wgts : ndarray
+        A 2D or 3D ndarray containing visibility weights
+        matching data shape. If None, and data is UVData,
+        will use UVData ~flags * nsample as weights.
+
+    axis : int or tuple
+        Axes along which to take sequential difference
+
+    Returns
+    -------
+    diff_data : ndarray or UVData object
+        A 2D or 3D complex differenced data array,
+        or UVData object holding differenced data
+        divided by sqrt(2).
+
+    prod_wgts : ndarray
+        If input data is ndarray, this is the geometric
+        mean of all wgts in the difference.
+    """
+    # type check
+    if isinstance(axis, (int, np.integer)):
+        axis = (axis,)
+    assert np.all([ax in [0, 1] for ax in axis]), "axis must be 0, 1 or both"
+
+    # perform differencing if data is an ndarray
+    if isinstance(data, np.ndarray):
+        # get weights
+        if wgts is None:
+            wgts = np.ones_like(data, dtype=np.float64)
+        else:
+            assert isinstance(wgts, np.ndarray), \
+                   "wgts must be ndarray if data is ndarray"
+
+        # difference data
+        for ax in axis:
+            data = (utils.dynamic_slice(data, slice(1, None), axis=ax) \
+                   - utils.dynamic_slice(data, slice(None, -1), axis=ax)) / np.sqrt(2)
+            wgts = np.sqrt(utils.dynamic_slice(wgts, slice(1, None), axis=ax) \
+                   * utils.dynamic_slice(wgts, slice(None, -1), axis=ax))
+
+        return data, wgts
+
+    # iterate over bls if UVData
+    elif isinstance(data, UVData):
+        # copy object
+        uvd = copy.deepcopy(data)
+
+        # get new time and freq arrays and UVData object
+        times = None
+        freqs = None
+        for ax in axis:
+            if ax == 0:
+                times = np.unique(uvd.time_array)[:-1]
+            elif ax == 1:
+                freqs = np.unique(uvd.freq_array)[:-1]
+        if times is not None or freqs is not None:
+            uvd.select(times=times, frequencies=freqs)
```

Comment:
Shouldn't the freqs and times be the average of the bins going into the subtraction?

Reply:
I think it could be either one depending on your interpretation: sort of like the different finite-differencing scheme.. One thing I don't like about changing the times and frequencies is that now they are mismatched w.r.t. the data we started with, which if possible I'd like to try to retain.